lib/util/cat.js

```diff
@@ -5,51 +5,272 @@
 const logger = require('../logger').child({ component: 'cat' })
 
-module.exports.handleCatHeaders = handleCatHeaders
-module.exports.parsedHeadersToTrans = parsedHeadersToTrans
+const HTTP_CAT_ID_HEADER = 'X-NewRelic-Id'
+const MQ_CAT_ID_HEADER = 'NewRelicID'
+const MATCH_CAT_ID_HEADER = new RegExp(
+  '^(?:' + HTTP_CAT_ID_HEADER + '|' + MQ_CAT_ID_HEADER + ')$',
+  'i'
+)
+const HTTP_CAT_TRANSACTION_HEADER = 'X-NewRelic-Transaction'
+const MQ_CAT_TRANSACTION_HEADER = 'NewRelicTransaction'
+const MATCH_CAT_TRANSACTION_HEADER = new RegExp(
+  '^(?:' + HTTP_CAT_TRANSACTION_HEADER + '|' + MQ_CAT_TRANSACTION_HEADER + ')$',
+  'i'
+)
+const HTTP_CAT_APP_DATA_HEADER = 'X-NewRelic-App-Data'
+const MQ_CAT_APP_DATA_HEADER = 'NewRelicAppData'
+const MATCH_CAT_APP_DATA_HEADER = new RegExp(
+  '^(?:' + HTTP_CAT_APP_DATA_HEADER + '|' + MQ_CAT_APP_DATA_HEADER + ')$',
+  'i'
+)
+
+/**
+ * Parses and decodes the CAT headers from incoming request and adds information to the active
+ * transaction
+ *
+ * @param {Object} headers incoming headers
+ * @param {string} encKey config.encoding_key used to decode CAT headers
+ * @param {Transaction} tx
+ */
+cat.handleCatHeaders = function handleCatHeaders(headers, encKey, tx) {
+  if (!encKey) {
+    logger.warn('Missing encoding key, not extract CAT headers!')
+    return
+  }
+
+  const { id, transactionId } = cat.extractCatHeaders(headers)
 
-function handleCatHeaders(incomingCatId, obfTransaction, encKey, transaction) {
   let parsedCatId = null
-  if (incomingCatId) {
-    parsedCatId = hashes.deobfuscateNameUsingKey(incomingCatId, encKey)
+  if (id) {
+    parsedCatId = hashes.deobfuscateNameUsingKey(id, encKey)
   }
 
   let externalTrans = null
-  if (obfTransaction) {
+  if (transactionId) {
     try {
-      externalTrans = JSON.parse(hashes.deobfuscateNameUsingKey(obfTransaction, encKey))
+      externalTrans = JSON.parse(hashes.deobfuscateNameUsingKey(transactionId, encKey))
     } catch (e) {
-      logger.trace('Got an unparsable CAT header x-newrelic-transaction: %s', obfTransaction)
+      logger.trace(`Got an unparsable CAT header ${HTTP_CAT_ID_HEADER} %s`, transactionId)
     }
   }
 
-  parsedHeadersToTrans(parsedCatId, externalTrans, transaction)
+  cat.parsedHeadersToTx(parsedCatId, externalTrans, tx)
+
+  if (tx.incomingCatId) {
+    logger.trace('Got inbound CAT headers in transaction %s from %s', tx.id, tx.incomingCatId)
+  }
+}
+
+/**
+ * Checks if hash is string
+ *
+ * @param {*} hash
+ * @return {Boolean}
+ */
+function _isValidReferringHash(hash) {
+  return typeof hash === 'string'
 }
 
-function parsedHeadersToTrans(parsedCatId, externalTrans, transaction) {
+/**
+ * Adds the appropriate keys to transaction based on the incoming parsed CAT data
+ *
+ * @param {string} parsedCatId decoded CAT id
+ * @param {Array} externalTrans CAT transaction
+ * @param {Transaction} tx active transaction
+ */
+cat.parsedHeadersToTx = function parsedHeadersToTx(parsedCatId, externalTrans, tx) {
```

Comment:
Can we just spell out 'Transaction'? I know its longer but I think spelling it out aids readability for future devs.